### DISCUSSION TITLE
Unable to get a valid stream when converting from gif to apng

### DISCUSSION THREAD

**Discussion by Kaoticz** [dlemstra/Magick.NET#1333] [Help] [answered]
I'm having trouble converting a stream from a `gif` file into a stream of an `apng` file.
I have the following method:
```cs
private static Stream GetAPngStreamAsync(Stream gifStream)
{
    using var imageCollection = new MagickImageCollection(gifStream, MagickFormat.Gif);
    var result = new MemoryStream();

    imageCollection.Write(result, MagickFormat.APng);

    return result;
}
```

Then I get the resulting stream and save it to a file using `File.WriteAllBytesAsync`.

```cs
using var processedStream = GetAPngStreamAsync(stream);
var buffer = new byte[processedStream.Length];

processedStream.Write(buffer);
await File.WriteAllBytesAsync("/home/user/Pictures/sample.png", buffer);
```
However, this always generates an invalid file that no image handling program can open.

But if I write to a file directly from `MagickImageCollection`, as in
```cs
imageCollection.Write("/home/user/Pictures/sample.png", MagickFormat.APng);
```
Then I get a perfectly functional file.

For my purposes though, I really need a valid stream and I'd like to avoid an I/O operation to write a new file, then read back from it.
It's very likely that I'm doing something silly here. Can I get some tips on how to generate a valid `apng` stream?

Also, this is unrelated, but how can I accelerate/slow down the animation in the resulting `apng`?

**Comment by dlemstra** [comment, marked as answer]
You will need to reset the position of the stream? You might also be worth to consider using `imageCollection.ToByteArray(MagickFormat.APng)` instead when you want to change the image into a byte array?

**Reply by Kaoticz** [reply]
That was it! `ToByteArray` worked flawlessly, thanks!

Now to seal the deal, how can I change the animation speed of the resulting `apng`?
I see that `IMagickImage<T>` has an `AnimationDelay` property, but it seems to only affect how long it takes for the whole animation to play again, not the speed which the animation frames are played.
